### PR TITLE
Remove EOL platforms Debian 8 and Ubuntu 14.04

### PR DESCRIPTION
**Pull request by ghoneycutt**

### CODE REVIEW

File: .nodeset.yml [hunk 1]
```diff
@@ -9,23 +9,3 @@ sets:
     nodes:
```
Comment (ghoneycutt) [inline .nodeset.yml:9]: What is this file used for? Can it go entirely?

**Comment by ghoneycutt** [comment]
There are other platforms that should be evaluated as they are not listed in the metadata and it is unclear as to which versions of the platform are even supported. Suggest officially removing support for these.

* FreeBSD
* OpenBSD
* Solaris
* ArchLinux
* Amazon

File: README.md [hunk 2]
```diff
@@ -367,7 +367,7 @@ This example shows how to install the MariaDB client and all of the bindings at
 Specify the version of the package you want with the `package_ensure` parameter.
 
 ```puppet
-class {'::mysql::client':
+class { 'mysql::client':
```
Comment (ghoneycutt) [inline README.md:370]: This file is just style updates I missed on the last pass

File: manifests/params.pp [hunk 3]
```diff
@@ -123,28 +123,24 @@
     'Suse': {
       case $::operatingsystem {
         'OpenSuSE': {
-          if versioncmp( $::operatingsystemmajrelease, '12' ) >= 0 {
```
Comment (ghoneycutt) [inline manifests/params.pp:126]: 12 is the smallest version we support (we support 12 and 15) so no need for the comparison.

File: manifests/params.pp [hunk 4]
```diff
@@ -153,31 +149,13 @@
       $config_file         = '/etc/my.cnf'
       $includedir          = '/etc/my.cnf.d'
       $datadir             = '/var/lib/mysql'
-      $log_error           = $::operatingsystem ? {
```
Comment (ghoneycutt) [inline manifests/params.pp:156]: Already a conditional above, so removing this pattern and adding to the correct block above.

File: manifests/params.pp [hunk 5]
```diff
@@ -188,34 +166,17 @@
...
-      $ruby_package_name   = $::operatingsystem ? {
-        /OpenSuSE/         => 'rubygem-mysql',
-        /(SLES|SLED)/      => 'ruby-mysql',
-      }
       $client_dev_package_name = 'libmysqlclient-devel'
       $daemon_dev_package_name = 'mysql-devel'
     }
 
     'Debian': {
-      if $::operatingsystem == 'Debian' and versioncmp($::operatingsystemrelease, '9') >= 0 {
```
Comment (ghoneycutt) [inline manifests/params.pp:200]: We support Debian 9 and 10 so all Debian is at least version 9 :)

File: .nodeset.yml (hunk 1 above)
Comment (pmcmaw) [reply .nodeset.yml:9]: This is just some old left over config that can be removed.

**Comment by pmcmaw** [comment]
Something to note as you will run into it is that we are currently facing issues testing Debian/Ubuntu. Which is why the Debian tests are not running and there are failures in your Ubuntu tests. 👎 It is something we have ticketed up but just haven't had the bandwidth to get around to yet, we are working on prioritizing and getting it worked on but cannot commit to a date for now. Issue is ticketed here: https://tickets.puppetlabs.com/browse/IAC-1595

**Comment by ghoneycutt** [comment]
@pmcmaw Does this mean this PR is on hold until that ticket is resolved or can we proceed with merging this?

What about these other unsupported platforms that are making the code more complex? Can we remove them?

File: README.md [hunk 6]
```diff
@@ -602,11 +602,11 @@ When the `/root/.mylogin.cnf` exists the environment variable `MYSQL_TEST_LOGIN_
 This is required if `create_root_user` and `create_root_login_file` are true. If `root_password` is 'UNSET', then `create_root_user` and `create_root_login_file` are assumed to be false --- that is, the MySQL root user and `/root/.mylogin.cnf` are not created.
 
 ```puppet
-class { '::mysql::server':
+class { 'mysql::server':
 root_password          => 'password',
 create_root_my_cnf     => false,
 create_root_login_file => true,
-login_file             => "puppet:///modules/${module_name}/mylogin.cnf",
+login_file             => 'puppet:///modules/mysql/mylogin.cnf',
```
Comment (alexjfisher) [inline README.md:609]: I don't think this change is correct.  I'm not a fan of `$module_name`, but here I think it refers to the `profile` module declaring `mysql::server` not to the `mysql` module itself.

**Comment by ghoneycutt** [comment]
The errors below are unrelated to this PR and were introduced with commit 8413dbeed33ff866360fba51f2a706f2367b931b

```
Pending: (Failures listed here are expected and do not affect your suite's status)

  1) mysql::password behaves like mysql::password function returns Sensitive with sensitive=true
     # should have a Returnvalue of Datatype Sensitive
     Failure/Error: is_expected.to run.with_params('password', true).and_return(sensitive('*2470C0C06DEE42FD1618BB99005ADCA2EC9D1E19'))
       expected mysql::password("password", true) to have returned Sensitive("*2470C0C06DEE42FD1618BB99005ADCA2EC9D1E19") instead of #<Sensitive [value redacted]>
     Shared Example Group: "mysql::password function" called from ./spec/functions/mysql_password_spec.rb:46
     # ./spec/functions/mysql_password_spec.rb:29:in `block (2 levels) in <top (required)>'
     # /opt/puppetlabs/pdk/share/cache/ruby/2.7.0/bin/rspec:23:in `load'
     # /opt/puppetlabs/pdk/share/cache/ruby/2.7.0/bin/rspec:23:in `<top (required)>'
     # /opt/puppetlabs/pdk/private/ruby/2.7.3/bin/bundle:23:in `load'
     # /opt/puppetlabs/pdk/private/ruby/2.7.3/bin/bundle:23:in `<main>'

  2) mysql::password non-namespaced shim mysql_password behaves like mysql::password function returns Sensitive with sensitive=true
     # should have a Returnvalue of Datatype Sensitive
     Failure/Error: is_expected.to run.with_params('password', true).and_return(sensitive('*2470C0C06DEE42FD1618BB99005ADCA2EC9D1E19'))
       expected mysql::password("password", true) to have returned Sensitive("*2470C0C06DEE42FD1618BB99005ADCA2EC9D1E19") instead of #<Sensitive [value redacted]>
     Shared Example Group: "mysql::password function" called from ./spec/functions/mysql_password_spec.rb:50
     # ./spec/functions/mysql_password_spec.rb:29:in `block (2 levels) in <top (required)>'
     # /opt/puppetlabs/pdk/share/cache/ruby/2.7.0/bin/rspec:23:in `load'
     # /opt/puppetlabs/pdk/share/cache/ruby/2.7.0/bin/rspec:23:in `<top (required)>'
     # /opt/puppetlabs/pdk/private/ruby/2.7.3/bin/bundle:23:in `load'
     # /opt/puppetlabs/pdk/private/ruby/2.7.3/bin/bundle:23:in `<main>'
```

**Comment by alexjfisher** [comment]
> There are other platforms that should be evaluated as they are not listed in the metadata and it is unclear as to which versions of the platform are even supported. Suggest officially removing support for these.
> 
> * FreeBSD
> * OpenBSD
> * Solaris
> * ArchLinux
> * Amazon

At least Solaris and Amazon Linux are listed as PE [supported agent platforms](https://puppet.com/docs/pe/2021.2/supported_operating_systems.html#supported_operating_systems_and_devices-supported-agent-platforms).

Dunno of the exact history though.  But I guess Solaris wasn't supported at the time of https://github.com/puppetlabs/puppetlabs-mysql/pull/774 ??

I'm also not keen on removing the unofficial support for non EOL OSes like ArchLinux.  I imagine @bastelfreak will be less keen still! :)

**Comment by pmcmaw** [comment]
@ghoneycutt we will proceed, we shouldn't let known issues stop the development of a module.

File: README.md (hunk 6 above)
Comment (ghoneycutt) [reply README.md]: Good eye @alexjfisher

I reverted this change.

**Comment by ghoneycutt** [comment]
@alexjfisher but what versions of Solaris and Amazon? :)

@pmcmaw This approach of Compatible and Supported is not great because it encourages baggage like this module has which makes maintaining it and refactoring it much more difficult and does not provide a way to test that compatibility even works.

All that said.. I wanted to get the conversation around this going and that is happening 🎉 

This PR though is ready for merge as it is just removing the EOL platforms Debian 8 and Ubuntu 14.04.

**Comment by pmcmaw** [comment]
> The errors below are unrelated to this PR and were introduced with commit [8413dbe](https://github.com/puppetlabs/puppetlabs-mysql/commit/8413dbeed33ff866360fba51f2a706f2367b931b)
> 
> ```
> Pending: (Failures listed here are expected and do not affect your suite's status)
> 
>   1) mysql::password behaves like mysql::password function returns Sensitive with sensitive=true
>      # should have a Returnvalue of Datatype Sensitive
>      Failure/Error: is_expected.to run.with_params('password', true).and_return(sensitive('*2470C0C06DEE42FD1618BB99005ADCA2EC9D1E19'))
>        expected mysql::password("password", true) to have returned Sensitive("*2470C0C06DEE42FD1618BB99005ADCA2EC9D1E19") instead of #<Sensitive [value redacted]>
>      Shared Example Group: "mysql::password function" called from ./spec/functions/mysql_password_spec.rb:46
>      # ./spec/functions/mysql_password_spec.rb:29:in `block (2 levels) in <top (required)>'
>      # /opt/puppetlabs/pdk/share/cache/ruby/2.7.0/bin/rspec:23:in `load'
>      # /opt/puppetlabs/pdk/share/cache/ruby/2.7.0/bin/rspec:23:in `<top (required)>'
>      # /opt/puppetlabs/pdk/private/ruby/2.7.3/bin/bundle:23:in `load'
>      # /opt/puppetlabs/pdk/private/ruby/2.7.3/bin/bundle:23:in `<main>'
> 
>   2) mysql::password non-namespaced shim mysql_password behaves like mysql::password function returns Sensitive with sensitive=true
>      # should have a Returnvalue of Datatype Sensitive
>      Failure/Error: is_expected.to run.with_params('password', true).and_return(sensitive('*2470C0C06DEE42FD1618BB99005ADCA2EC9D1E19'))
>        expected mysql::password("password", true) to have returned Sensitive("*2470C0C06DEE42FD1618BB99005ADCA2EC9D1E19") instead of #<Sensitive [value redacted]>
>      Shared Example Group: "mysql::password function" called from ./spec/functions/mysql_password_spec.rb:50
>      # ./spec/functions/mysql_password_spec.rb:29:in `block (2 levels) in <top (required)>'
>      # /opt/puppetlabs/pdk/share/cache/ruby/2.7.0/bin/rspec:23:in `load'
>      # /opt/puppetlabs/pdk/share/cache/ruby/2.7.0/bin/rspec:23:in `<top (required)>'
>      # /opt/puppetlabs/pdk/private/ruby/2.7.3/bin/bundle:23:in `load'
>      # /opt/puppetlabs/pdk/private/ruby/2.7.3/bin/bundle:23:in `<main>'
> ```

These are fine, they are actually marked as pending so wont affect the test suite.

**Comment by ghoneycutt** [comment]
@pmcmaw The spec tests are failing here instead of being pending as they are locally. The other failing tests are the known issues. We good to merge?

**Comment by pmcmaw** [comment]
Would it be possible to remove Ubuntu20 from the metadata.json as currently we don't want it running on CI with the known issues that are currently raised against it.

Im unsure as to why the other Ubuntu OSs are currently failing in this PR as they are passing against main in the nightly run. If you maybe try rebasing? Last nights nightly run: https://github.com/puppetlabs/puppetlabs-mysql/actions/runs/999406897

**Comment by ghoneycutt** [comment]
@pmcmaw I removed Ubuntu 20 from the metadata as you asked and rebased.

**Comment by pmcmaw** [comment]
Thanks @ghoneycutt I was hoping that rebasing sorted your Ubuntu errors. But it hasn't and last nights nightly ran green on main therefore it suggests there is somewhere in the logic something is going wrong. 👎 Im unable to merge until all of the above OSs are running green.

**Comment by ghoneycutt** [comment]
@pmcmaw This appears to be an issue with the test environment.

From local testing.

```
Finished in 7 minutes 14 seconds (files took 45.96 seconds to load)
111 examples, 0 failures


pid 13059 exit 0
Successful on 1 nodes: ["localhost:2222, litmusimage/ubuntu:18.04"]
Checking connectivity for ["localhost:2222"]
Connectivity check PASSED.
```

**Comment by ghoneycutt** [comment]
@pmcmaw all green :)

**Review by pmcmaw — APPROVED** [review]

**Comment by pmcmaw** [comment]
Im happy with the changes from my end, @alexjfisher just wanting to check if you happy enough for me to merge as I can see you commented earlier in the thread :-)

**Comment by ghoneycutt** [comment]
Alex was just commenting on my proposal for removing OS's that are not officially supported. This is ready for merge.

**Pull request merged**